orphaning_tree.go

```diff
@@ -43,24 +43,29 @@ func (tree *orphaningTree) unorphan(hash []byte) {
 	tree.ndb.Unorphan(hash)
 }
 
-// Save the underlying Tree. Saves orphans too.
-func (tree *orphaningTree) SaveVersion(version int64) {
-	// Save the current tree at the given version. For each saved node, we
-	// delete any existing orphan entries in the previous trees.
-	// This is necessary because sometimes tree re-balancing causes nodes to be
-	// incorrectly marked as orphaned, since tree patterns after a re-balance
-	// may mirror previous tree patterns, with matching hashes.
-	tree.ndb.SaveBranch(tree.root, func(node *Node) {
-		// The node version is set here since it isn't known until we save.
-		// Note that we only want to set the version for inner nodes the first
-		// time, as they represent the beginning of the lifetime of that node.
-		// So unless it's a leaf node, we only update version when it's 0.
-		if node.version == 0 || node.isLeaf() {
-			node.version = version
-		}
-		tree.unorphan(node._hash())
-	})
-	tree.ndb.SaveOrphans(version, tree.orphans)
+// SaveAs saves the underlying Tree and assigns it a new version.
+// Saves orphans too.
+func (tree *orphaningTree) SaveAs(version int64) {
+	tree.version = version
```

Comment:
Lets assert here that version = tree.version + 1 before doing this?  (e.g. panic otherwise).  I can imagine that in the future we might want to be able to load an outdated version and resume from there, but for now I don't think the assertion should fail.

(BTW, If *Tree is never assigned, then it can just be Tree.  Since the containing struct is always used as a pointer, golang lets orphaningTree.Tree be a pointer receiver.)

Reply:
As I was about to do this, I realised we can make a simplification now by getting rid of latestVersion, and then we set the version in *VersionedTree. Check it out and let me know if you still think we need to check (v' == v+1): https://github.com/tendermint/iavl/commit/4e423126e41a9b650eeb4890a769bd8061beed5d

Reply:
I think now the semantics of latestVersion changes.  See my comments... 
What does it meant to load a previous version of the tree, and then call Hash or LatestVersion?  Maybe there should be LatestHash() and Hash() as well as LatestVersion() and Version()?